content/posts/jspi-and-function-pointer-cast-handling.md

```diff
@@ -0,0 +1,380 @@
+with extra arguments. The WebAssembly `call_indirect` function which calls a
+function pointer traps if the function pointer does not have the expected type.
+For example suppose we have code like this:
+```C
+typedef int (*F)(int, int);
+
+int handler0(int x, int y) {
+  logString("Handler 0 fetch x");
+  int resx = awaitFakeFetch(x);
+  logString("      ... fetch y");
+  int resy = awaitFakeFetch(y);
+  return x * y;
+}
+
+int handler1(int x) {
+  logString("Handler 1 fetch x");
+  int res = awaitFakeFetch(x);
+  return res * res;
+}
+
+F handlers[] = {(F)handler0, (F)handler1 };
+
+WASM_EXPORT("fakePyFunc")
+void fakePyFunc(int func_index, int x, int y) {
+  int res = handlers[func_index](x, y);
+  logString("Got result:");
+  logInt(res);
+}
+```
+When we call `fakePyFunc()` with `func_index` set to `0` it works, but when set
+to `1`, we get a crash:
+```js
+RuntimeError: null function or function signature mismatch
+```
+See the complete example
+[here](https://github.com/hoodmane/jspi-blog-examples/tree/main/7-fpcast).
+
+Calls from JavaScript into WebAssembly ignore extra arguments, so we had been
+dealing with this situation before by adding a JavaScript trampoline:
+```js
+const imports = {
+  env: {
+    ...
+    fpcastTrampoline(fptr, x, y) {
+      return functionPointerTable.get(fptr)(x, y);
+    },
+    ...
+  },
+};
+```
+and then in C invoking the handler as follows:
+```C
+  fpcastTrampoline(handlers[func_index], x, y);
+```
+This fixes the function pointer casts but when we attempt to stack switch it
+raises:
+```js
+SuspendError: trying to suspend JS frames
+```
+See the complete example
+[here](https://github.com/hoodmane/jspi-blog-examples/tree/main/8-js-frame).
+
+## Handling function pointer casts without JavaScript
+
+Luckily, `wasm-gc` has added a `ref.test` instruction which can be used to check
+the signature of a function before calling it. We can use this to count how many
+arguments the function was actually defined with and cast it to the right
+signature before making the call. Here is an example wat module
+
+<style>
+pre { line-height: 125%; }
+td.linenos .normal { color: inherit; background-color: transparent; padding-left: 5px; padding-right: 5px; }
+span.linenos { color: inherit; background-color: transparent; padding-left: 5px; padding-right: 5px; }
+td.linenos .special { color: #000000; background-color: #ffffc0; padding-left: 5px; padding-right: 5px; }
+span.linenos.special { color: #000000; background-color: #ffffc0; padding-left: 5px; padding-right: 5px; }
+
+.pygments {
+  font-size: .78em;
+}
+
+.highlight .hll { background-color: #49483e }
+.highlight  { background: #272822; color: #f8f8f2 }
+.highlight .c { color: #75715e } /* Comment */
+.highlight .err { color: #960050; background-color: #1e0010 } /* Error */
+.highlight .k { color: #66d9ef } /* Keyword */
+.highlight .l { color: #ae81ff } /* Literal */
+.highlight .n { color: #f8f8f2 } /* Name */
+.highlight .o { color: #f92672 } /* Operator */
```

Comment:
So sad that the markdown library does not support syntax highlighting for wat...